crates/matrix-sdk/src/send_queue.rs

```diff
@@ -746,6 +787,162 @@ impl QueueStorage {
+                if let Some(event_id) = de.event_id {
+                    // Has been sent; send a redaction.
+                    let room = client
+                        .get_room(&self.room_id)
+                        .ok_or(RoomSendQueueError::RoomDisappeared)?;
+
+                    // Ideally we'd use the send queue to send the redaction, but the protocol has
+                    // changed the shape of a room.redaction after v11, so keep it simple and try
+                    // once here.
+
+                    if let Err(err) = room.redact(&event_id, None, None).await {
+                        warn!("error when sending a redact for {event_id}: {err}");
+                    }
+                } else {
+                    // Still local (sending must have failed); redact the local echo.
+                    let removed = store
+                        .remove_send_queue_event(&self.room_id, &de.transaction_id)
+                        .await
+                        .map_err(RoomSendQueueStorageError::StorageError)?;
+
+                    if !removed {
+                        warn!("missing local echo upon dependent redact??");
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    #[instrument(skip(self))]
+    async fn apply_dependent_events(&self) -> Result<(), RoomSendQueueError> {
+        // Keep the lock until we're done touching the storage.
+        let _being_sent = self.being_sent.read().await;
+
+        let client = self.client()?;
+        let store = client.store();
+
+        let dependent_events = store
+            .list_dependent_send_queue_events(&self.room_id)
+            .await
+            .map_err(RoomSendQueueStorageError::StorageError)?;
+
+        let num_initial_dependent_events = dependent_events.len();
+
+        let canonicalized_dependent_events = canonicalize_dependent_events(&dependent_events);
+        let mut num_dependent_events = canonicalized_dependent_events.len();
+
+        debug!(
+            num_dependent_events,
+            num_initial_dependent_events, "starting handling of dependent events"
+        );
+
+        for de in canonicalized_dependent_events {
+            let dependent_event_id = de.id;
+
+            match self.try_apply_single_dependent_event(&client, de).await {
+                Ok(()) => {
+                    // The dependent event has been successfully applied, forget about it.
+                    store
+                        .remove_dependent_send_queue_event(&self.room_id, dependent_event_id)
+                        .await
+                        .map_err(RoomSendQueueStorageError::StorageError)?;
+
+                    num_dependent_events -= 1;
+                }
+
+                Err(err) => {
+                    warn!("error when applying single dependent event: {err}");
+                }
+            }
+        }
+
+        debug!(num_dependent_events, "stopped handling dependent events");
```

Comment:
Perhaps calling this `leftover_dependent_events` would make it clearer what this is about.